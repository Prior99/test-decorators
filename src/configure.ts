export type It = (name: string, fn: ItFunction) => void
export type ItFunction = (done?: () => void) => Promise<void> | void

export type Describe = (name: string, fn: DescribeFunction) => void
export type DescribeFunction = () => Promise<void> | void

export interface ConfigureOptions {
    it: It,
    itOnly?: It,
    describe: Describe,
    describeOnly?: Describe
}

export let configuration: ConfigureOptions

export function configure(options: ConfigureOptions) {
    Object.assign(configuration, options)
}

declare var global: any

export function initializeConfiguration() {
    // Shamelessly adapted from https://github.com/purposeindustries/window-or-global
    // istanbul ignore next
    const globalOrWindow = (typeof self === "object" && self.self === self && self) ||
        (typeof global === "object" && global.global === global && global)

    configuration = {
        it: (globalOrWindow as any).it,
        describe: (globalOrWindow as any).describe,
        describeOnly: (globalOrWindow as any).describe.only,
        itOnly: (globalOrWindow as any).it.only,
    }
}

initializeConfiguration()
