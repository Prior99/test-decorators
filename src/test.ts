import "reflect-metadata"
import { TestOptionsInput, parseTestOptions, isTestOptionsInput } from "./options"
import { configuration } from "./configure"

export type TestFunction = (instance: any) => void

export function getTests(target: any): TestFunction[] {
    const tests = Reflect.getMetadata(`test-decorator:tests`, target)
    if (typeof tests !== "undefined") { return tests }
    const newTests: TestFunction[] = []
    Reflect.defineMetadata(`test-decorator:tests`, newTests, target)
    return newTests
}

export type TestDecorator =
    (target: Object, property: string | symbol, descriptor: PropertyDescriptor) => PropertyDescriptor

export function test<Params>(
    target: Object,
    property: string | symbol,
    descriptor: PropertyDescriptor,
): PropertyDescriptor
export function test<Params>(options?: TestOptionsInput<Params>): TestDecorator
/**
 * Decorate a method with `@test` or `@test("name of the test")` or @test({ ... })` to
 * automatically have it run as a test inside an `it` call.
 *
 * @param options Can be provided or left out, so the decorator can be called with `@test()` or
 *                like `@test`. The name of the test or a configuration object. @see TestOptions
 *
 * @return The decorated method.
 */
export function test<Params>(
    arg1: Object | TestOptionsInput<Params>,
    arg2?: string | symbol,
    arg3?: PropertyDescriptor,
) {
    // Load the `it` and `itOnly` functions from the configuration which is stored singleton.
    // It might have been configured to overwrite the mocha ones.
    const { it, itOnly } = configuration
    if (typeof it !== "function") {
        throw new Error("'it' not found. Did you call 'configure'?")
    }
    const options = isTestOptionsInput(arg1) ? parseTestOptions(arg1) : {
        name: () => (arg2 as string),
        params: undefined,
        only: false,
    }
    const { name, params, only } = options
    const decorator: TestDecorator = (target, property, descriptor) => {
        const tests = getTests(target.constructor)
        const itFunction = only ? itOnly : it
        if (only && typeof itOnly !== "function") {
            throw new Error("Call to 'itOnly' occured but it was not configured.")
        }
        // If `params` was not specified, call `it` once with no parameters.
        if (!params) {
            tests.push(instance => {
                // prevents super-class from executing the tests of extending classes
                if (!(instance instanceof target.constructor)) return;
                itFunction(name(), async (...args: any[]) => await descriptor.value.apply(instance, args))
            })
            return descriptor
        }
        // Call `it` once for all parameters configured in `params`.
        tests.push(...params.map(param => {
            return (instance: any) => {
                // prevents super-class from executing the tests of a child class
                // while also preventing the child class from executing parent tests
                if (!(instance.constructor.name === target.constructor.name)) return;
                itFunction(
                    name(param),
                    async (...args: any[]) => await descriptor.value.apply(instance, [param, ...args]),
                )
            }
        }))
        return descriptor
    }
    // Function was used as a decorator with a configuration (`@test(...)`).
    if (isTestOptionsInput(arg1)) {
        return decorator
    }
    // Function was used as a decorator without a configuration (`@test`).
    return decorator(arg1, arg2, arg3)
}
