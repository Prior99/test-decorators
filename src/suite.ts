import "reflect-metadata"
import { SuiteOptionsInput, parseSuiteOptions, isSuiteOptionsInput, SuiteOptions } from "./options"
import { configuration } from "./configure"
import { getTests } from "./test"

export type Constructable<T> = {
    new(): T;
}

export function suite(options?: SuiteOptionsInput): ClassDecorator
export function suite<U, T extends Constructable<U>>(Ctor?: T): T
export function suite<U, T extends Constructable<U>>(arg1: T | SuiteOptionsInput) {
    const { describe, describeOnly } = configuration
    if (typeof describe !== "function") {
        throw new Error("'describe' not found. Did you call 'configure'?")
    }
    const decorator: ClassDecorator = Ctor => {
        let options: SuiteOptions
        if (isSuiteOptionsInput(arg1)) {
            options = parseSuiteOptions(arg1)
        } else {
            options = { name: Ctor.name }
        }
        const { name, only } = options
        const describeFunction = only ? describeOnly : describe
        if (only && typeof describeOnly !== "function") {
            throw new Error("Call to 'describeOnly' occured but it was not configured.")
        }
        describeFunction(name, () => {
            const instance = new (Ctor as any)()
            const tests = getTests(instance.constructor)
            tests.forEach(test => test(instance))
        })
    }
    if (isSuiteOptionsInput(arg1)) {
        return decorator
    }
    return decorator(arg1)
}
