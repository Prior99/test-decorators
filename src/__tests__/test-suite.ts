import { test as testDecorator } from "../test"
import { suite } from "../suite"
import { configure, It, Describe } from "../configure"

describe("`suite`", () => {
    let mockIt: jest.Mock<It>
    let mockItOnly: jest.Mock<It>
    let mockDescribe: jest.Mock<Describe>
    let mockDescribeOnly: jest.Mock<Describe>

    beforeEach(() => {
        mockIt = jest.fn()
        mockItOnly = jest.fn()
        mockDescribe = jest.fn()
        mockDescribeOnly = jest.fn()
        configure({
            it: mockIt,
            itOnly: mockItOnly,
            describe: mockDescribe,
            describeOnly: mockDescribeOnly,
        })
    })

    test("without a configuration", () => {
        @suite
        class A {
            @testDecorator
            private testSomething() { return }
        }
        expect(mockDescribe.mock.calls).toMatchSnapshot()
    })

    test("with `describe` not being configured", () => {
        configure({ it, describe: undefined })
        expect(() => {
            @suite
            class A {
                @testDecorator
                private testSomething() { return }
            }
        }).toThrowErrorMatchingSnapshot()
    })

    test("with `only` set to `true` and `describeOnly` not being configured", () => {
        configure({ it, describe, describeOnly: undefined })
        expect(() => {
            @suite({ only: true })
            class A {
                @testDecorator
                private testSomething() { return }
            }
        }).toThrowErrorMatchingSnapshot()
    })

    test("the constructor is called", () => {
        const mockConstructorImpl = jest.fn()
        @suite
        class A {
            constructor() {
                mockConstructorImpl()
            }
        }
        mockDescribe.mock.calls[0][1]()
        expect(mockConstructorImpl).toHaveBeenCalledTimes(1)
    })

    test("parameterized tests are called with `this` set to a valid instance of the decorated class", () => {
        const mockTestImpl = jest.fn()
        @suite
        class A {
            @testDecorator
            public someTest() { mockTestImpl(this) }
        }
        expect(mockDescribe).toHaveBeenCalledTimes(1)
        mockDescribe.mock.calls[0][1]()
        expect(mockIt).toHaveBeenCalledTimes(1)
        mockIt.mock.calls[0][1]()
        expect(mockTestImpl).toHaveBeenCalledTimes(1)
        expect(mockTestImpl.mock.calls[0][0].constructor).toBe(A)
    })
})
