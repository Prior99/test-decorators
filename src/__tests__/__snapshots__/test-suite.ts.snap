// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`suite\` with \`describe\` not being configured 1`] = `"'describe' not found. Did you call 'configure'?"`;

exports[`\`suite\` with \`only\` set to \`true\` and \`describeOnly\` not being configured 1`] = `"Call to 'describeOnly' occured but it was not configured."`;

exports[`\`suite\` without a configuration 1`] = `
Array [
  Array [
    "A",
    [Function],
  ],
]
`;
